README
Enables you to automatically click the left mouse button.

For compilation/installation instructions, see INSTALL.

Website: http://xautoclick.sourceforge.net/



website/news.txt

<B>NEWS</B>
<P>

<div class="news">
unreleased
<p>
* Remove fltk, gtk1, and qt3 support
</div>

<DIV CLASS="news">
v0.31 is released. <P>
Changes in this version:
<P>
* fix compiler check in configure<BR>
* fix missing include (QT4 gui)<BR>
* fix linking of FLTK version<BR>
<BR>
<I>Sun Apr  1 18:36:15 CEST 2012 --Ivo</I>
</DIV>

<P>

<DIV CLASS="news">
v0.30 is released. <P>
After four years, some new code has been added again! Changes in this version:
<P>
* Major code cleanup<br>
* New Qt 4 GUI<br>
* New FLTK GUI<br>
<BR>
<I>Mon May 24 17:53:21 CEST 2010 --Ivo</I>
</DIV>

<P>

<DIV CLASS="news">
v0.20 is released. <P>
If v0.19 worked correctly for you, there's no need to
upgrade. The only things changed are various typos and a little more verbose
output of configure in case none of the needed development packages are
installed. This confused a lot of new users, hence the update.
<P>
If it does not work for you, please file a bugreport on the mailinglist.
Contrary to what some seem to believe, this project is not abandoned. It's just
that the current code seems to work correctly and I never get any bugreports or
feature requests :)
<BR>
<I>Fri Oct 17 14:04:48 CEST 2008 --Ivo</I>
</DIV>

<P>

<DIV CLASS="news">
v0.19 is released. I migrated the source repository to Subversion, increased
the upper limits on all values and fixed a small bug in the Qt3 version.
<BR>
<I>Tue Aug 15 17:14:32 CEST 2006 --Ivo</I>
</DIV>

<P>

<DIV CLASS="news">
v0.18 is released. Please test on as many X11 platforms you can find and
if it doesn't work, send a <A HREF="mailinglists.html">bug report</A>.
<BR>
<I>Tue Mar 28 17:11:56 CEST 2006 --Ivo</I>
</DIV>

<P>

<DIV CLASS="news">
Project started. Initial import in CVS and set up this very basic website.<BR>
<I>Sat Mar 18 18:53:16 CET 2006 --Ivo</I>
</DIV>

<P>

<DIV CLASS="news">
Project approved by SourceForge staff.<BR>
<I>Fri Mar 17 2006 --Ivo</I>
</DIV>



src/qt4.cpp
/* xAutoClick, Qt 4 GUI
 *
 * Copyright (C) 2010 Ivo van Poorten
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include "qt4.h"

extern "C" {
#include "main.h"
#include "osdep.h"
}

#include <X11/Xlib.h>
#include <X11/extensions/XTest.h>

static Display *display;
static QApplication *app;
static MyWidget *mywidget;

void click_mouse_button(void) {
    XTestFakeButtonEvent(display, 1, True, CurrentTime);
    XTestFakeButtonEvent(display, 1, False, CurrentTime);
    XFlush(display);
}

void set_alarm(int ms) {
    mywidget->timer->start(ms);
}

int get_spin_value(spin_t spin) {
    return mywidget->spins[spin]->value();
}

void set_spin_value(spin_t spin, int value) {
    mywidget->spins[spin]->setValue(value);
}

void set_button_sensitive(button_t button, int state) {
    mywidget->buttons[button]->setEnabled(state);
}

void MyWidget::tap(void) {
    common_tap_button();
}

void MyWidget::stop(void) {
    common_stop_button();
}

void MyWidget::start(void) {
    common_start_button();
}

void MyWidget::timer_done(void) {
    common_alarm_callback();
}

MyWidget::MyWidget(QWidget *parent) : QWidget(parent) {

    struct foo {
        QString label;
        int min, max, val;
    } bar[4] = {
        { "Pre-delay",   1, 1<<30, 2000 }, { "Interval",    1, 1<<30, 1000 },
        { "Random +/-",  1, 10240,   64 }, { "# of clicks", 1, 10240,   32 }
    };
    QString butnames[3] = { "Tap", "Stop", "Start" };

    QVBoxLayout *vbox = new QVBoxLayout;

    for (int c=0; c<4; c++) {
        QHBoxLayout *layout = new QHBoxLayout;
        spins[c] = new QSpinBox;
        spins[c]->setMaximum(bar[c].max);
        spins[c]->setMinimum(bar[c].min);
        spins[c]->setValue(bar[c].val);

        layout->addWidget(new QLabel(bar[c].label));
        layout->addWidget(spins[c]);
        vbox->addLayout(layout);
    }

    QHBoxLayout *layout = new QHBoxLayout;

    for (int c=0; c<3; c++)
        layout->addWidget(buttons[c] = new QPushButton(butnames[c]));

    connect(buttons[0], SIGNAL(clicked()), this, SLOT(tap(void)));
    connect(buttons[1], SIGNAL(clicked()), this, SLOT(stop(void)));
    connect(buttons[2], SIGNAL(clicked()), this, SLOT(start(void)));

    vbox->addLayout(layout);
    setWindowTitle("Qt4AutoClick");
    setLayout(vbox);

    timer = new QTimer;
    timer->setSingleShot(TRUE);
    connect(timer, SIGNAL(timeout()), this, SLOT(timer_done()));
}

int init_gui(int argc, char **argv) {
    if (!(display = XOpenDisplay(NULL))) {
        fprintf(stderr, "Unable to open X display\n");
        return 0;
    }

    app = new QApplication(display);
    mywidget = new MyWidget;

    return 1;
}

void close_gui(void) {
}

void main_loop(void) {
    mywidget->show();
    app->exec();
}


xautoclick/guiascii.c
/* xAutoClick -- ASCII GUI
 *
 * Copyright (C) 2006 Ivo van Poorten
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <stdio.h>
#include <X11/Xlib.h>
#include <X11/extensions/XTest.h>

#include "main.h"
#include "osdep.h"

static Display *display;
static int predelay, interval, randomfactor, numberofclicks, sleeptime;

void help(void) {
    printf("aautoclick\n\n");
    printf("p - set pre-delay\n");
    printf("h - help\n");
    printf("i - set interval\n");
    printf("r - set random +/-\n");
    printf("n - set number of clicks\n");
    printf("t - tap\n");
    printf("s - start\n");
    printf("q - quit\n");
    printf("\n");
}

void click_mouse_button(void) {
    XTestFakeButtonEvent(display, 1, True, CurrentTime);
    XTestFakeButtonEvent(display, 1, False, CurrentTime);
    XFlush(display);
}

void set_alarm(int ms) {
    sleeptime = ms;
}

int get_spin_value(spin_t spin) {

    switch(spin) {
    case SPIN_PREDELAY:
        return predelay;
        break;
    case SPIN_INTERVAL:
        return interval;
        break;
    case SPIN_RANDOM:
        return randomfactor;
        break;
    case SPIN_NUMBER:
        return numberofclicks;
        break;
    }

    return 0;
}

void set_spin_value(spin_t spin, int value) {

    switch(spin) {
    case SPIN_PREDELAY:     predelay       = value;     break;
    case SPIN_INTERVAL:     interval       = value;     break;
    case SPIN_RANDOM:       randomfactor   = value;     break;
    case SPIN_NUMBER:       numberofclicks = value;     break;
    }
}

void set_button_sensitive(button_t button, int state) {
}

int init_gui(int argc, char **argv) {

    if (!(display = XOpenDisplay(NULL))) {
        fprintf(stderr, "Unable to open X display\n");
        return 0;
    }

    predelay       = 2000;
    interval       = 1024;
    randomfactor   = 64;
    numberofclicks = 32;

    help();
    return 1;
}

void close_gui(void) {
    XCloseDisplay(display);
}

static void print_variables(void) {
    printf("pre-delay:          %i\n", predelay);
    printf("interval:           %i\n", interval);
    printf("random +/-:         %i\n", randomfactor);
    printf("number of clicks:   %i\n", numberofclicks);
    printf("\n");
}

#define is_space(x) (x==' ' || x=='\t')
#define is_eol(x) (x=='\n')

static void flush_to_eol(int c) {
    if (is_eol(c)) return;
    while(!is_eol(c)) { c = fgetc(stdin); }
}

static void read_int(int *ptr) {
    printf("new value: ");
    fscanf(stdin, "%i", ptr);
    flush_to_eol(0);
}

void main_loop(void) {
    int c;

    print_variables();
    printf("> ");
    c = fgetc(stdin);

    while (c) {

        while(is_space(c)) { c = fgetc(stdin); }
        if (is_eol(c)) {
            printf("> ");
            c = fgetc(stdin);
            continue;
        }

        flush_to_eol(c);

        switch (c) {
        case 'p':   read_int(&predelay);            break;
        case 'i':   read_int(&interval);            break;
        case 'r':   read_int(&randomfactor);        break;
        case 'n':   read_int(&numberofclicks);      break;
        case 't':   common_tap_button();            break;
        case 'h':   help();                         break;
        case 's':   goto run;                       break;
        case 'q':   goto quit_out;                  break;
        default:    printf("unknown command\n");    break;
        }

        print_variables();

        usleep(100 * 1000);
        printf("> ");
        c = fgetc(stdin);
    }

run:
    common_start_button();

    while (sleeptime) {
        usleep(sleeptime * 1000);
        sleeptime = 0;
        common_alarm_callback();
    }

quit_out:
    return;
}


src/guigtk2.c
/* xAutoClick -- GTK1 GUI
 *
 * Copyright (C) 2006 Ivo van Poorten
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#include <stdio.h>
#include <X11/Xlib.h>
#include <X11/extensions/XTest.h>
#include <gtk/gtk.h>

#include "main.h"

static Display *display;

static GtkWidget *gAutoClick, *tap_button, *stop_button, *start_button;
static GtkWidget *predelay_spin, *interval_spin, *random_spin;
static GtkWidget *totalclicks_spin;

void click_mouse_button(void) {
    XTestFakeButtonEvent(display, 1, True, CurrentTime);
    XTestFakeButtonEvent(display, 1, False, CurrentTime);
    XFlush(display);
}

static gboolean myalarm(gpointer data) {
    common_alarm_callback();
    return FALSE;   /* call only once; FALSE destroys timer */
}

void set_alarm(int ms) {
    g_timeout_add(ms, myalarm, NULL);
}

int get_spin_value(spin_t spin) {

    switch(spin) {
    case SPIN_PREDELAY:
        return gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(predelay_spin));
        break;
    case SPIN_INTERVAL:
        return gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(interval_spin));
        break;
    case SPIN_RANDOM:
        return gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(random_spin));
        break;
    case SPIN_NUMBER:
        return gtk_spin_button_get_value_as_int(GTK_SPIN_BUTTON(totalclicks_spin));
        break;
    }

    return 0;
}

void set_spin_value(spin_t spin, int value) {

    switch(spin) {
    case SPIN_PREDELAY:
        break;
    case SPIN_INTERVAL:
        gtk_spin_button_set_value(GTK_SPIN_BUTTON(interval_spin), value);
        break;
    case SPIN_RANDOM:
        gtk_spin_button_set_value(GTK_SPIN_BUTTON(random_spin), value);
        break;
    case SPIN_NUMBER:
        break;
    }
}

void set_button_sensitive(button_t button, int state) {

    switch(button) {
    case BUTTON_TAP:
        gtk_widget_set_sensitive(tap_button, state ? TRUE : FALSE);
        break;
    case BUTTON_STOP:
        gtk_widget_set_sensitive(stop_button, state ? TRUE : FALSE);
        break;
    case BUTTON_START:
        gtk_widget_set_sensitive(start_button, state ? TRUE : FALSE);
        break;
    }
}

static void on_tap_button_clicked(GtkButton *button, gpointer user_data) {
    common_tap_button();
}

static void on_stop_button_clicked(GtkButton *button, gpointer user_data) {
    common_stop_button();
}

static void on_start_button_clicked(GtkButton *button, gpointer user_data) {
    common_start_button();
}

static GtkWidget *create_gAutoClick(void) {
  GtkWidget *gAutoClick;
  GtkWidget *vbox1;
  GtkWidget *hbox1;
  GtkWidget *predelay_label;
  GtkWidget *label8;
  GtkObject *predelay_spin_adj;
  GtkWidget *hbox2;
  GtkWidget *interval_label;
  GtkWidget *label9;
  GtkObject *interval_spin_adj;
  GtkWidget *hbox3;
  GtkWidget *random_label;
  GtkWidget *label10;
  GtkObject *random_spin_adj;
  GtkWidget *hbox5;
  GtkWidget *totalclicks_label;
  GtkWidget *label11;
  GtkObject *totalclicks_spin_adj;
  GtkWidget *hbox4;

  gAutoClick = gtk_window_new (GTK_WINDOW_TOPLEVEL);
  gtk_object_set_data (GTK_OBJECT (gAutoClick), "gAutoClick", gAutoClick);
  gtk_container_set_border_width (GTK_CONTAINER (gAutoClick), 4);
  gtk_window_set_title (GTK_WINDOW (gAutoClick), "gAutoClick");
/*  gtk_window_set_position (GTK_WINDOW (gAutoClick), GTK_WIN_POS_CENTER); */
  gtk_window_set_policy (GTK_WINDOW (gAutoClick), FALSE, FALSE, FALSE);

  vbox1 = gtk_vbox_new (FALSE, 0);
  gtk_widget_ref (vbox1);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "vbox1", vbox1,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (vbox1);
  gtk_container_add (GTK_CONTAINER (gAutoClick), vbox1);

  hbox1 = gtk_hbox_new (FALSE, 0);
  gtk_widget_ref (hbox1);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "hbox1", hbox1,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (hbox1);
  gtk_box_pack_start (GTK_BOX (vbox1), hbox1, TRUE, TRUE, 0);

  predelay_label = gtk_label_new ("Pre-delay");
  gtk_widget_ref (predelay_label);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "predelay_label", predelay_label,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (predelay_label);
  gtk_box_pack_start (GTK_BOX (hbox1), predelay_label, FALSE, FALSE, 0);
  gtk_label_set_justify (GTK_LABEL (predelay_label), GTK_JUSTIFY_FILL);

  label8 = gtk_label_new ("");
  gtk_widget_ref (label8);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "label8", label8,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (label8);
  gtk_box_pack_start (GTK_BOX (hbox1), label8, TRUE, FALSE, 0);

  predelay_spin_adj = gtk_adjustment_new (2000, 0, 1<<30, 1, 10, 0);
  predelay_spin = gtk_spin_button_new (GTK_ADJUSTMENT (predelay_spin_adj), 1, 0);
  gtk_widget_ref (predelay_spin);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "predelay_spin", predelay_spin,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (predelay_spin);
  gtk_box_pack_start (GTK_BOX (hbox1), predelay_spin, FALSE, FALSE, 0);
  gtk_widget_set_size_request (predelay_spin, 64, -2);
  gtk_spin_button_set_numeric (GTK_SPIN_BUTTON (predelay_spin), TRUE);

  hbox2 = gtk_hbox_new (FALSE, 0);
  gtk_widget_ref (hbox2);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "hbox2", hbox2,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (hbox2);
  gtk_box_pack_start (GTK_BOX (vbox1), hbox2, TRUE, TRUE, 0);

  interval_label = gtk_label_new ("Interval");
  gtk_widget_ref (interval_label);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "interval_label", interval_label,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (interval_label);
  gtk_box_pack_start (GTK_BOX (hbox2), interval_label, FALSE, FALSE, 0);

  label9 = gtk_label_new ("");
  gtk_widget_ref (label9);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "label9", label9,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (label9);
  gtk_box_pack_start (GTK_BOX (hbox2), label9, TRUE, FALSE, 0);

  interval_spin_adj = gtk_adjustment_new (1000, 0, 1<<30, 1, 10, 0);
  interval_spin = gtk_spin_button_new (GTK_ADJUSTMENT (interval_spin_adj), 1, 0);
  gtk_widget_ref (interval_spin);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "interval_spin", interval_spin,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (interval_spin);
  gtk_box_pack_start (GTK_BOX (hbox2), interval_spin, FALSE, FALSE, 0);
  gtk_widget_set_size_request (interval_spin, 64, -2);
  gtk_spin_button_set_numeric (GTK_SPIN_BUTTON (interval_spin), TRUE);

  hbox3 = gtk_hbox_new (FALSE, 0);
  gtk_widget_ref (hbox3);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "hbox3", hbox3,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (hbox3);
  gtk_box_pack_start (GTK_BOX (vbox1), hbox3, TRUE, TRUE, 0);

  random_label = gtk_label_new ("Random +/-");
  gtk_widget_ref (random_label);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "random_label", random_label,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (random_label);
  gtk_box_pack_start (GTK_BOX (hbox3), random_label, FALSE, FALSE, 0);

  label10 = gtk_label_new ("");
  gtk_widget_ref (label10);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "label10", label10,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (label10);
  gtk_box_pack_start (GTK_BOX (hbox3), label10, TRUE, FALSE, 0);

  random_spin_adj = gtk_adjustment_new (250, 0, 10240, 1, 10, 0);
  random_spin = gtk_spin_button_new (GTK_ADJUSTMENT (random_spin_adj), 1, 0);
  gtk_widget_ref (random_spin);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "random_spin", random_spin,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (random_spin);
  gtk_box_pack_start (GTK_BOX (hbox3), random_spin, FALSE, FALSE, 0);
  gtk_widget_set_size_request (random_spin, 64, -2);
  gtk_spin_button_set_numeric (GTK_SPIN_BUTTON (random_spin), TRUE);

  hbox5 = gtk_hbox_new (FALSE, 0);
  gtk_widget_ref (hbox5);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "hbox5", hbox5,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (hbox5);
  gtk_box_pack_start (GTK_BOX (vbox1), hbox5, TRUE, TRUE, 0);

  totalclicks_label = gtk_label_new ("Total clicks");
  gtk_widget_ref (totalclicks_label);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "totalclicks_label", totalclicks_label,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (totalclicks_label);
  gtk_box_pack_start (GTK_BOX (hbox5), totalclicks_label, FALSE, FALSE, 0);

  label11 = gtk_label_new ("");
  gtk_widget_ref (label11);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "label11", label11,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (label11);
  gtk_box_pack_start (GTK_BOX (hbox5), label11, TRUE, FALSE, 0);

  totalclicks_spin_adj = gtk_adjustment_new (32, 1, 10240, 1, 10, 0);
  totalclicks_spin = gtk_spin_button_new (GTK_ADJUSTMENT (totalclicks_spin_adj), 1, 0);
  gtk_widget_ref (totalclicks_spin);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "totalclicks_spin", totalclicks_spin,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (totalclicks_spin);
  gtk_box_pack_start (GTK_BOX (hbox5), totalclicks_spin, FALSE, FALSE, 0);
  gtk_widget_set_size_request (totalclicks_spin, 64, -2);
  gtk_spin_button_set_numeric (GTK_SPIN_BUTTON (totalclicks_spin), TRUE);

  hbox4 = gtk_hbox_new (FALSE, 0);
  gtk_widget_ref (hbox4);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "hbox4", hbox4,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (hbox4);
  gtk_box_pack_start (GTK_BOX (vbox1), hbox4, TRUE, TRUE, 0);

  tap_button = gtk_button_new_with_label ("     Tap     ");
  gtk_widget_ref (tap_button);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "tap_button", tap_button,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (tap_button);
  gtk_box_pack_start (GTK_BOX (hbox4), tap_button, TRUE, TRUE, 0);
  gtk_container_set_border_width (GTK_CONTAINER (tap_button), 2);

  stop_button = gtk_button_new_with_label ("Stop");
  gtk_widget_ref (stop_button);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "stop_button", stop_button,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (stop_button);
  gtk_box_pack_start (GTK_BOX (hbox4), stop_button, TRUE, TRUE, 0);
  gtk_container_set_border_width (GTK_CONTAINER (stop_button), 2);

  start_button = gtk_button_new_with_label ("Start");
  gtk_widget_ref (start_button);
  gtk_object_set_data_full (GTK_OBJECT (gAutoClick), "start_button", start_button,
                            (GtkDestroyNotify) gtk_widget_unref);
  gtk_widget_show (start_button);
  gtk_box_pack_start (GTK_BOX (hbox4), start_button, TRUE, TRUE, 0);
  gtk_container_set_border_width (GTK_CONTAINER (start_button), 2);

  gtk_signal_connect (GTK_OBJECT (gAutoClick), "delete_event",
                      GTK_SIGNAL_FUNC (gtk_main_quit),
                      NULL);
  gtk_signal_connect (GTK_OBJECT (tap_button), "clicked",
                      GTK_SIGNAL_FUNC (on_tap_button_clicked),
                      NULL);
  gtk_signal_connect (GTK_OBJECT (stop_button), "clicked",
                      GTK_SIGNAL_FUNC (on_stop_button_clicked),
                      NULL);
  gtk_signal_connect (GTK_OBJECT (start_button), "clicked",
                      GTK_SIGNAL_FUNC (on_start_button_clicked),
                      NULL);

  return gAutoClick;
}

int init_gui(int argc, char **argv) {

    if (!(display = XOpenDisplay(NULL))) {
        fprintf(stderr, "Unable to open X display\n");
        return 0;
    }

    gtk_set_locale();
    gtk_init(NULL, NULL);

    gAutoClick = create_gAutoClick();
    gtk_widget_show(gAutoClick);

    return 1;
}

void close_gui(void) {
    XCloseDisplay(display);
}

void main_loop(void) {
    gtk_main();
}
